ncd-calculator/src/components/KGridNCD.tsx

```diff
@@ -0,0 +1,303 @@
+import React, {useCallback, useEffect, useState} from "react";
+import {
+    Block,
+    createInitialState,
+    GridObject,
+    GridState,
+    isInSelectedBlock,
+    Position, selectRandomBlock, swapBlocks
+} from "@/services/kgrid.ts";
+
+const GridCell: React.FC<{
+    object: GridObject,
+    position: Position,
+    onSelect: (pos: Position) => void,
+    isSelected: boolean
+}> = ({ object, position, onSelect, isSelected }) => {
+    const displayContent = object.content.length > 20
+        ? object.content.substring(0, 20) + '...'
+        : object.content;
+
+    return (
+        <div
+            className={`grid-cell ${isSelected ? 'selected' : ''}`}
+            onClick={() => onSelect(position)}
+            title={object.content}
+        >
+            <div className="cell-content">
+                {displayContent}
+            </div>
+        </div>
+    );
+};
+
+export const KGridNCD: React.FC<{
+    width?: number,
+    height?: number,
+    objects?: GridObject[]
+}> = ({
+          width = 3,
+          height = 3,
+          objects = [
+              // Default sample objects if none provided
+              { id: "0", content: "Lions are large cats that live in Africa." },
+              { id: "1", content: "Tigers are striped cats from Asia." },
+              { id: "2", content: "Dogs are loyal pets with keen senses." },
+              { id: "3", content: "Wolves are wild canines that live in packs." },
+              { id: "4", content: "Dolphins are intelligent marine mammals." },
+              { id: "5", content: "Sharks are predatory fish with cartilage skeletons." },
+              { id: "6", content: "Eagles are birds of prey with sharp vision." },
+              { id: "7", content: "Penguins are flightless birds that swim." },
+              { id: "8", content: "Frogs are amphibians that start life in water." }
+          ]
+      }) => {
+    // State hooks
+    const [gridState, setGridState] = useState<GridState>(() =>
+        createInitialState(width, height, objects)
+    );
+    const [selectedBlock1, setSelectedBlock1] = useState<Block | null>(null);
+    const [selectedBlock2, setSelectedBlock2] = useState<Block | null>(null);
+    const [isOptimizing, setIsOptimizing] = useState<boolean>(false);
+    const [optimizationSteps, setOptimizationSteps] = useState<number>(0);
+    const [horizontalReflection, setHorizontalReflection] = useState<boolean>(false);
```

Comment:
we don't need gui for mutations. they are generated randomly by computer as part of greedy hill climb or simulated annealing optimization algo